composer/callbacks/memory_snapshot.py

```diff
@@ -0,0 +1,161 @@
+            ... )
+
+    .. note::
+        Memory snapshot is only supported for GPU devices.
+
+    Args:
+        skip_batches (int, optional): Number of batches to skip before starting recording memory snapshot. Defaults to 1.
+        interval (Union[int, str, Time], optional): Time string specifying how long to record the tensor allocation.
+            For example, ``interval='3ba'`` means 3 batches are recorded. Default: '3ba'.
+        max_entries (int, optional): Maximum number of memory alloc/free events to record. Defaults to 100000.
+        folder (str, optional): A format string describing the folder containing the memory snapshot files.
+            Defaults to ``'{{run_name}}/torch_traces'``.
+        filename (str, optional): A format string describing how to name the memory snapshot files.
+            Defaults to ``'rank{{rank}}.{{batch}}.pickle'``.
+        remote_file_name (str, optional): A format string for the memory snapshot remote file name.
+            Defaults to ``'{{run_name}}/torch_traces/rank{{rank}}.{{batch}}.pickle'``.
+
+            Whenever a trace file is saved, it is also uploaded as a file according to this format string.
+            The same format variables as for ``filename`` are available.
+
+            .. seealso:: :doc:`Uploading Files</trainer/file_uploading>` for notes for file uploading.
+
+            Leading slashes (``'/'``) will be stripped.
+
+            To disable uploading trace files, set this parameter to ``None``.
+        overwrite (bool, optional): Whether to override existing memory snapshots. Defaults to False.
+
+            If False, then the trace folder as determined by ``folder`` must be empty.
+    """
+
+    def __init__(
+        self,
+        skip_batches: int = 1,
+        interval: Union[int, str, Time] = '3ba',
+        max_entries: int = 100000,
+        folder: str = '{run_name}/torch_traces',
+        filename: str = 'rank{rank}.{batch}.pickle',
+        remote_file_name: Optional[str] = '{run_name}/torch_traces/rank{rank}.{batch}.pickle',
+        overwrite: bool = False,
+    ) -> None:
+        self.batches_left_to_skip = skip_batches
+        # Check that the interval timestring is parsable and convert into time object
+        self.interval = Time.from_input(interval, TimeUnit.BATCH)
+        self.max_entries = max_entries
+        self.folder = folder
+        self.folder_name = None
+        self.filename = filename
+        self.remote_file_name = remote_file_name
+        self.overwrite = overwrite
+        self._enabled = True
+        self._start_time = None
+
+    def init(self, state: State, logger: Logger) -> None:
+        # Not relying on `torch.cuda.is_available()` since the model could be on CPU.
+        model_device = next(state.model.parameters()).device
+
+        if model_device.type not in ('cuda', 'meta'):
+            warnings.warn(f'The memory snapshot only works on CUDA devices, but the model is on {model_device.type}.')
```

Comment:
I prefer we raise an exception so the user explicitly knows they won't get memory profiles.  Lmk what you  all think cc: @mvpatel2000

Reply:
I saw the check here https://github.com/mosaicml/composer/blob/dev/composer/callbacks/memory_monitor.py#L89-L94 and took it.

Reply:
I think it's fine to emit warning -- there are weird scenarios where user manually moves around model and callback is still useful

Reply:
Sounds good w/me